tests/integration/test_types_integration.py

```diff
@@ -200,6 +201,119 @@ def test_digest(trino_connection):
+    query_time_with_timezone(trino_connection, '-08:00')
+    query_time_with_timezone(trino_connection, '+08:00')
+    query_time_with_timezone(trino_connection, '+05:30')
+
+
+def query_time_with_timezone(trino_connection, timezone_str):
+    tz = datetime.datetime.strptime(timezone_str, "%z").tzinfo
+    time_0 = datetime.time(1, 23, 45, 0).replace(tzinfo=tz)
+    time_3 = datetime.time(1, 23, 45, 123000).replace(tzinfo=tz)
+    time_6 = datetime.time(1, 23, 45, 123456).replace(tzinfo=tz)
+
+    SqlTest(trino_connection) \
+        .add_field(sql="CAST(null AS TIME WITH TIME ZONE)", python=None) \
+        .add_field(sql="CAST(null AS TIME(0) WITH TIME ZONE)", python=None) \
+        .add_field(sql="CAST(null AS TIME(3) WITH TIME ZONE)", python=None) \
+        .add_field(sql="CAST(null AS TIME(6) WITH TIME ZONE)", python=None) \
+        .add_field(sql="CAST(null AS TIME(9) WITH TIME ZONE)", python=None) \
+        .add_field(sql="CAST(null AS TIME(12) WITH TIME ZONE)", python=None) \
+        .add_field(sql="CAST('01:23:45 %s' AS TIME(0) WITH TIME ZONE)" % (timezone_str), python=time_0) \
+        .add_field(sql="TIME '01:23:45.123 %s'" % (timezone_str), python=time_3) \
+        .add_field(sql="CAST('01:23:45.123 %s' AS TIME(3) WITH TIME ZONE)" % (timezone_str), python=time_3) \
+        .add_field(sql="CAST('01:23:45.123456 %s' AS TIME(6) WITH TIME ZONE)" % (timezone_str), python=time_6) \
+        .add_field(sql="CAST('01:23:45.123456789 %s' AS TIME(9) WITH TIME ZONE)" % (timezone_str), python=time_6) \
+        .add_field(sql="CAST('01:23:45.123456789123 %s' AS TIME(12) WITH TIME ZONE)" % (timezone_str), python=time_6) \
+        .execute()
+
+
+def test_timestamp(trino_connection):
+    timestamp_0 = datetime.datetime(2001, 8, 22, 1, 23, 45, 0)
+    timestamp_3 = datetime.datetime(2001, 8, 22, 1, 23, 45, 123000)
+    timestamp_6 = datetime.datetime(2001, 8, 22, 1, 23, 45, 123456)
+
+    SqlTest(trino_connection) \
+        .add_field(sql="CAST(null AS TIMESTAMP)", python=None) \
+        .add_field(sql="CAST(null AS TIMESTAMP(0))", python=None) \
+        .add_field(sql="CAST(null AS TIMESTAMP(3))", python=None) \
+        .add_field(sql="CAST(null AS TIMESTAMP(6))", python=None) \
+        .add_field(sql="CAST(null AS TIMESTAMP(9))", python=None) \
+        .add_field(sql="CAST(null AS TIMESTAMP(12))", python=None) \
+        .add_field(sql="CAST('2001-08-22 01:23:45' AS TIMESTAMP(0))", python=timestamp_0) \
+        .add_field(sql="TIMESTAMP '2001-08-22 01:23:45.123'", python=timestamp_3) \
+        .add_field(sql="CAST('2001-08-22 01:23:45.123' AS TIMESTAMP(3))", python=timestamp_3) \
+        .add_field(sql="CAST('2001-08-22 01:23:45.123456' AS TIMESTAMP(6))", python=timestamp_6) \
+        .add_field(sql="CAST('2001-08-22 01:23:45.123456789' AS TIMESTAMP(9))", python=timestamp_6) \
+        .add_field(sql="CAST('2001-08-22 01:23:45.123456789123' AS TIMESTAMP(12))", python=timestamp_6) \
```

Comment:
If we're actually rounding this should have failed since value read back would be `.123457` instead.